Issue 5562: Wrong outstanding appends calculation in AppendBatchSizeTracker using connection pooling

RaulGracia opened the pull request:
**Change log description**  
Fixed the calculation of append batches in the client if connection pooling is enabled and multiple flows are multiplexed over the same connection.

**Purpose of the change**  
Fixes #5562.
Fixes #5531.

**What the code does**  

To solve #5562, this PR advocates to keep 1 `AppendBatchSizeTracker` object per flow. To this end:

- We introduce a new class (i.e., `FlowToBatchSizeTracker`), to keep the mapping between flows and batch trackers. It is based on `SimpleCache`, which provides us with thread-safety and an eviction mechanism, so this mapping will not grow forever.
- As `AppendBatchSizeTracker` is required at the `TCPClientConnection` and `CommandEncoder` levels, is there where we instantiate `FlowToBatchSizeTracker`. In this sense, this PR also removes the useless mapping of flow ids and batch trackers at the `FlowHandler` level, which in turn fixes #5531.
- Note that both `CommandEncoder` and `TCPClientConnection` fetch the appropriate `AppendBatchSizeTracker` by using `Flow.toFlowID(requestId)`. That is, the `AppendBatchSizeTracker` object is associated to a flow (i.e., most significant bits), irrespective of its sequence number (i.e., least significant bits).
- There is some minor and non-functional cleanup in `CommandEncoder` (e.g., wrong javadoc, redundant castings).

**How to verify it**  
Added new tests for `FlowToBatchSizeTracker`. I have tested this PR on AWS with connection pooling enabled (hundreds of segments and writers) and results are better in terms of latency.

shrids commented on client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:
```diff
@@ -115,7 +113,7 @@ public void run() {
                     if (command instanceof WireCommands.DataAppended) {
                         WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
-                        batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                        flowToBatchSizeTracker.getAppendBatchSizeTrackerByFlowId(dataAppended.getRequestId()).recordAck(dataAppended.getEventNumber());
```
Using `io.pravega.client.connection.impl.Flow#toFlowID` api to get the requestId will ensure  only the flow id is used. This will ensure requests via the RawClient do not create multiple `AppendBatchSizeTrackers`.

RaulGracia replied:
Done.

shrids commented on client/src/main/java/io/pravega/client/connection/impl/FlowToBatchSizeTracker.java:
```diff
@@ -0,0 +1,69 @@
+    /**
+     * Maximum time that the want to keep an inactive flow entry in the cache.
+     */
+    private static final Duration TRACKER_CACHE_EXPIRATION_TIME = Duration.ofSeconds(60);
```
This value can be same as `io.pravega.client.connection.impl.TcpClientConnection#SOCKET_TIMEOUT_MS`

RaulGracia replied:
Done.

shrids reviewed (commented):
Great Find 👍 
The changes LGTM apart from a couple of questions...

shrids reviewed (approved):
LGTM

shrids commented on client/src/main/java/io/pravega/client/connection/impl/TcpClientConnection.java:
```diff
@@ -115,7 +113,7 @@ public void run() {
                     if (command instanceof WireCommands.DataAppended) {
                         WireCommands.DataAppended dataAppended = (WireCommands.DataAppended) command;
-                        batchSizeTracker.recordAck(dataAppended.getEventNumber());
+                        flowToBatchSizeTracker.getAppendBatchSizeTrackerByFlowId(dataAppended.getRequestId()).recordAck(dataAppended.getEventNumber());
```
A response to a conditionalAppend would create a AppendBatchSizetracker instance...

RaulGracia replied:
Thanks @shrids, that's correct, but then we need to use `Flow.toFlowID()` in `CommandEncoder` as well as I added in the last commit. Otherwise, the `CommandEncoder` and `TcpClientConnection` refer to different keys for the same `AppendBatchSizeTracker`.

tkaitchuck commented on client/src/main/java/io/pravega/client/connection/impl/FlowHandler.java:
```diff
@@ -185,18 +147,6 @@ public void process(Reply cmd) {
-            if (batchSizeTracker != null) {
-                long pendingAckCount = batchSizeTracker.recordAck(dataAppended.getEventNumber());
-                // Only publish client side metrics when there is some metrics notifier configured for efficiency.
-                if (!metricNotifier.equals(MetricNotifier.NO_OP_METRIC_NOTIFIER)) {
-                    metricNotifier.updateSuccessMetric(CLIENT_OUTSTANDING_APPEND_COUNT, writerTags(dataAppended.getWriterId().toString()),
-                            pendingAckCount);
-                }
```
Are we loosing this metric?

RaulGracia replied:
We have never used it since it was introduced as far as I know, and it may reduce the computation overhead of the client when processing the ack. We can bring it back in the future if necessary.

shrids reviewed (approved):
+1

fpj commented on client/src/main/java/io/pravega/client/connection/impl/FlowToBatchSizeTracker.java:
```diff
@@ -0,0 +1,69 @@
+    @Getter(value = AccessLevel.PACKAGE)
+    private final SimpleCache<Integer, AppendBatchSizeTracker> flowToBatchSizeTrackerMap;
+
+    public FlowToBatchSizeTracker() {
```
Should this constructor and this class be package protected rather than public?

RaulGracia replied:
Done.

fpj commented on client/src/main/java/io/pravega/client/connection/impl/FlowToBatchSizeTracker.java:
```diff
@@ -0,0 +1,69 @@
+     */
+    public AppendBatchSizeTracker getAppendBatchSizeTrackerByFlowId(int flowId) {
+        if (flowToBatchSizeTrackerMap.putIfAbsent(flowId, new AppendBatchSizeTrackerImpl()) == null) {
+            log.info("Instantiating new batch sze tracker for flow: {}", flowId);
```
This is a debug message.

RaulGracia replied:
Done.

fpj reviewed (changes requested):
A couple of minor points.

leapsky commented:
Experiments with `5000 segments, 250MBps, 1KB events` and `connection pooling enabled`, `1` and `100 writers` ran successfully.
Also, I ran experiments with `1000 segments, 250MBps, 1KB events` and `connection pooling disabled` which showed similar results that we had for master.

RaulGracia commented:
@fpj thanks for the feedback, I have addressed your comments. If you don't have any further comments, please try to give a go to this PR as it would be important to cherry-pick it to `r0.9`.
